schema/representations.md

```diff
@@ -0,0 +1,256 @@
+
+Kinds and their Representations
+-------------------------------
+
+A type at the Schema level must be mapped onto a representation expressible
+within the Data Model.
+
+For all Schema kinds which have equivalents at the Data Model level, there is
+a default representation.  The default representation this is, broadly speaking,
+a no-op: a string remains a string, and a map is easy to enough to represent as,
+well, a map!
+
+For the perceived kinds (those introduced at the Schema level, with no direct
+equivalent in the Data Model), most also have default mappings: for example,
+structs will simply be mapped into maps; enum values are mapped into strings.
+Unions are a special case: there's no single widely agreed upon transcription
+pattern for unions from before the establishment of IPLD, so these require
+an explicit choice.
+
+Customized representation strategies are also possible.
+Each Schema type definition may specify such a custom representation, overriding
+the default.
+
+Representation strategies are identified by one primary keyword (which indicates
+what Kind in the Data Model will be transcribed to), and may contain several
+more detailed parameters which are specific to the strategy (e.g., the struct
+map representation has a concept of 'default' values which is unique to that
+strategy).
+
+### table
+
+(Kinds for which no meaningful customization is possible -- Null, Boolean,
+Integer, Float, String, Bytes, and List -- are elided.
+N.B., it's not ruled out that custom representations may be introduced for
+these kinds in the future; there are simply none to date.
+Link is also not described in this section, as it's a rather unique business.)
+
+- Map
+	- `map` representation -- the default -- transcribes to `map` in the Data Model.
+	- `stringpairs` representation -- transcribes to `string` in the Data Model.
```

Comment:
I think @mikeal has a visceral reaction to the mention of `stringpairs` previously. Is this something we're seeing in the wild or is it just a might-be-nice-to-have?

Reply:
so, i see why having these is good in that it expresses what can be done by the schema beyond basic type bindings. my primary concern is just opening the door to “every feature ever” at this layer. like, how many of these “string*” features will we add to the spec and what is the reason for allowing at this level or not?

Reply:
> my primary concern is just opening the door to “every feature ever” at this layer.

100%.  It's a super legitimate concern.  There's clear reasons for "some" features of this form.  It would also clearly be extremely problematic if this got out of hand with too many extensions.  Every new member in this set of known representations adds implementation work to every single IPLD library that will ever be written -- and if we were to add new members after the first stable versions are released, will create partial compatibility tables between libraries (!!) -- and that should not be taken lightly.

I don't know how we police future choices other than always being clear that the discussion starts at "default: No." and require any more additions make a very strong case for themselves.

One rule of thumb that might help is "if it seems turing complete, hard No".  Similarly, "if the parser would be recursive, hard No".  I might suggest "if it requires bittwiddling or math, lean against"; "if this seems like a one-off for someone's pet project, No"; "if there aren't existing of examples of this pattern done in the wild, No"... perhaps the biggest one might be "Does this enable other features that aren't otherwise possible, or have radically different performance characteristics if carried out by the library?"    This last question is why we're here in the first place (the answer is "yes" because it lets us have complex map key types, which was otherwise intractable).

Beyond that, discretion and caution.  And (rapidly) increasing reticence over time, due to that partial compatibility concern.

Reply:
That kind of thought process would be good to capture somewhere, not just because it'd be amusing to see you try and codify that into documentation form; these are the kinds of details that get lost with people-churn so boundaries start to slip because nobody knows where they should be.

Reply:
_engages holodeck poker room_

I'll see your preemptive schadenfreude, and raise you one epistemological dread!

 . . .

More seriously, though, yes, we should try to capture that somewhere; thoughts on how to draft that would indeed by welcome.